download/oci_dowload.go

```diff
@@ -0,0 +1,366 @@
+			if err != nil {
+				retry++
+			} else {
+				retry = 0
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *OCIDownload) oneShot(ctx context.Context) error {
+	m := metrics.New()
+	resp, err := d.download(ctx, m)
+	if err != nil {
+		if d.f != nil {
+			d.f(ctx, Update{ETag: "", Bundle: nil, Error: err, Metrics: m, Raw: nil})
+		}
+		return err
+	}
+
+	d.longPollingEnabled = resp.longPoll
+
+	if d.f != nil {
+		d.f(ctx, Update{ETag: resp.etag, Bundle: resp.b, Error: nil, Metrics: m, Raw: resp.raw})
+	}
+	return nil
+}
+
+func (d *OCIDownload) download(ctx context.Context, m metrics.Metrics) (*downloaderResponse, error) {
+	d.logger.Debug("OCI - Download starting.")
+
+	preferences := []string{fmt.Sprintf("modes=%v,%v", defaultBundleMode, deltaBundleMode)}
+	if d.longPollingEnabled && d.config.Polling.LongPollingTimeoutSeconds != nil {
+		wait := fmt.Sprintf("wait=%s", strconv.FormatInt(*d.config.Polling.LongPollingTimeoutSeconds, 10))
+		preferences = append(preferences, wait)
+
+		// fetch existing response header timeout value on the http client's transport and
+		// clear it for the long poll request
+		current := d.client.Config().ResponseHeaderTimeoutSeconds
+		if *current != 0 {
+			d.respHdrTimeoutSec = *current
+			t := int64(0)
+			d.client = d.client.SetResponseHeaderTimeout(&t)
+		}
+	}
+	preferValue := fmt.Sprintf("%v", strings.Join(preferences, ";"))
+	d.client = d.client.WithHeader("Prefer", preferValue)
+
+	m.Timer(metrics.BundleRequest).Start()
+	_, layers, err := d.pull(d.path)
+	if err != nil {
+		return &downloaderResponse{}, fmt.Errorf("failed to pull %s: %w", d.path, err)
+	}
+	//currently it has 3 as it has the manifest, tar and config layers
+	if len(layers) != 3 {
+		return nil, fmt.Errorf("expected 3 layers only")
+	}
+
+	tarballDescriptor := ocispec.Descriptor{}
+	for i := range layers {
+		if layers[i].MediaType == "application/vnd.oci.image.layer.v1.tar+gzip" {
+			tarballDescriptor = layers[i]
+			break
+		}
+	}
+	if tarballDescriptor.MediaType == "" {
+		return nil, fmt.Errorf("no tarball descriptor found in the layers")
+	}
+
+	bundleFilePath := filepath.Join(d.localStorePath, "blobs", "sha256", string(tarballDescriptor.Digest.Hex()))
+	fileReader, err := os.Open(bundleFilePath)
+	if err != nil {
+		return nil, err
+	}
+	loader := bundle.NewTarballLoaderWithBaseURL(fileReader, d.localStorePath)
+	reader := bundle.NewCustomReader(loader).WithBaseDir(d.localStorePath)
+	bundleInfo, err := reader.Read()
+	if err != nil {
+		return &downloaderResponse{}, fmt.Errorf("unexpected error %w", err)
+	}
+
+	m.Timer(metrics.BundleRequest).Stop()
+
+	return &downloaderResponse{
+		b:        &bundleInfo,
+		raw:      fileReader,
+		etag:     "",
+		longPoll: false,
+	}, nil
+}
+
+func (d *OCIDownload) pull(ref string) (*ocispec.Descriptor, []ocispec.Descriptor, error) {
+	clientConfig := d.client.Config()
+	authHeader := make(http.Header)
+	resolver := docker.NewResolver(docker.ResolverOptions{
```

Comment:
There's a lot of stuff going on here... is it all about opa-as-client authenticating with different registries?

Reply:
I did a bit of refactoring in the latest commit here so it can be easier to read. I hope that clears things up.